### DISCUSSION TITLE
Debugging with Picoprobe/pico-debug, OpenOCD, and GDB

### DISCUSSION THREAD

**Discussion by approxdec** [earlephilhower/arduino-pico#1243] [General]
g'day.

under this heading on the Code page ...

what is the FULL command for pico-debug?

should this work on windows 11?

thanks

joe

**Comment by maxgerhardt** [comment]
Pico Probe and pico-debug are two very different things. One is an external hardware debugger the other one is a program loaded in cache (XIP flash cache) or RAM to enable debugging on the same board. 

Which one would you like to use?
By "FULL command" do you mean the OpenOCD command for connecting to these debuggers on the commandline?

**Comment by approxdec** [comment]
g'day.
thanks.
i refer to pico-debug the external hardware debugger
and I mean   the OpenOCD command for connecting to this debugger  on the
commandline

joe

**Reply by maxgerhardt** [reply]
..But [Pico-Debug](https://github.com/majbthrd/pico-debug/) is the software and not an *external* hardware debugger. That's the Picoprobe. Well anyways, ignoring the naming confusion here, to connect to a Picoprobe external hardware debugger, flash [the latest Picoprobe firmware](https://github.com/raspberrypi/picoprobe/releases) on it (BOOTSEL mode + drag+drop the .uf2), connect the Picoprobe do the target Pico [as documented](https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf) ("Picoprobe Wiring"), [grab OpenOCD for your OS](https://github.com/earlephilhower/pico-quick-toolchain/releases/tag/1.5.0-b) and assuming you're in the root of the extract directory you can do

```
bin\openocd.exe -f share/openocd/scripts -f interface/cmsis-dap.cfg -f target/rp2040.cfg
```

**Comment by approxdec** [comment]
yes.
muddling names.
I refer to this in fact.

https://www.raspberrypi.com/documentation/microcontrollers/debug-probe.html

and thanks for line you gave me.
On Windows 11 that *should *work for the picoprobe but it doesn't. Windows
11 will not recognise the device, can't find it, whatever.

But what about the debug probe?

Do you have possible command line for that?

All this seems to have been worked out for Linux but it does not transfer
to Windows.

thanks for your help.

I got to the git page from the arduino-pico package install.

regards
Joe

**Reply by maxgerhardt** [reply]
>But what about the debug probe?

The debug probe you linked runs the same Picoprobe firmware. There's no difference in how it should enumerate on a PC.

What does the device manager or Zadig show in regars to devices when you plug in the Picoprobe / Raspnerry Pi Debug Probe?

Also, please let's keep this conversation in one message thread. You're always creating new threads when you reply by mail. All responses could be nicely beneath in one thread.

**Reply by approxdec** [reply]
Well I'm sorry about that but it is midnight here and I'm on a mobile phone.
Device manager shows the UART and cmsis as ports but everything I have tried from reading does not work.
now Max. I am going to get some sleep which means I won't be around for 6-7 hours. [Australia, Sydney time]

**Reply by approxdec** [reply]
goodmorning, good evening.
this morning i have 'forgotten' what i have done before
plugged the debug probe and device manager shows
![image](https://user-images.githubusercontent.com/3886404/222549398-f93d66b1-ac51-48b1-ac39-4204e5d15a47.png)
so i do
```
F:\openocd-pico-debug>openocd -f scripts\interface\cmsis-dap.cfg -f scripts\target\rp2040.cfg
Open On-Chip Debugger 0.12.0-g4257276 (2023-01-27-10:19)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
scripts\target\rp2040.cfg:9: Error: Can't find target/swj-dp.tcl
in procedure 'script'
at file "embedded:startup.tcl", line 28
at file "scripts\target\rp2040.cfg", line 9

F:\openocd-pico-debug>
```
so - i do not have the debug probe connected to a pico.
is openocd looking for a debug probe connection to a pico?

and there is the next problem
[https://www.raspberrypi.com/documentation/microcontrollers/debug-probe.html](url)
the docs are not very helpful for the SWCLK GND SWDIO connection to the pico.
i am using female lead ends to a 3 pin male header on the pico.

probably not, but anyway

more searching and reading

joe

**Reply by approxdec** [reply]
if i change line 9 in rp2040.cfg to

source [find scripts/target/swj-dp.tcl]

i get this

```
F:\openocd-pico-debug>openocd -f scripts\interface\cmsis-dap.cfg -f scripts\target\rp2040.cfg
Open On-Chip Debugger 0.12.0-g4257276 (2023-01-27-10:19)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Warn : An adapter speed is not selected in the init scripts. OpenOCD will try to run the adapter at the low speed (100 kHz)
Warn : To remove this warnings and achieve reasonable communication speed with the target, set "adapter speed" or "jtag_rclk" in the init scripts.
Error: unable to find a matching CMSIS-DAP device

so i am back to the cmsis-dap device problem again.

joe
```

**Reply by maxgerhardt** [reply]
The COM22 for the CMSIS-DAP adapter does not look good at all. Looks like it has USB CDC drivers loaded instead of WinUSB. Can you use https://zadig.akeo.ie/ to change the drivers fro the CMSIS-DAP (Interface 0) to WinUSB and retry?

**Reply by approxdec** [reply]
working on it

**Reply by approxdec** [reply]
```
F:\openocd-pico-debug>openocd -f scripts\interface\cmsis-dap.cfg -f scripts\target\rp2040.cfg
Open On-Chip Debugger 0.12.0-g4257276 (2023-01-27-10:19)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Warn : An adapter speed is not selected in the init scripts. OpenOCD will try to run the adapter at the low speed (100 kHz)
Warn : To remove this warnings and achieve reasonable communication speed with the target, set "adapter speed" or "jtag_rclk" in the init scripts.
Info : Using CMSIS-DAPv2 interface with VID:PID=0x2e8a:0x000c, serial=E6614103E76B342F
Info : CMSIS-DAP: SWD supported
Info : CMSIS-DAP: Atomic commands supported
Info : CMSIS-DAP: Test domain timer supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 100 kHz
Error: Failed to connect multidrop rp2040.dap0
```
device manager sees the uart - interface 1 com21

zadig = cmsis-dap = WinUSB
CDC-ACM UART interface 1 = usbser v1.0.0.0

**Reply by maxgerhardt** [reply]
Alright looks much better, if now at least finds the debug adapter. So that's one down. It now seems that the Pickprobe is not correctly connected to the target board. Can you take a picture of the wireup between the debugger and the debuggee?

**Comment by approxdec** [comment]
ok i am working with this
```
F:\openocd-pico-debug>openocd -f scripts\interface\cmsis-dap.cfg -f scripts\target\rp2040.cfg
Open On-Chip Debugger 0.12.0-g4257276 (2023-01-27-10:19)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Warn : An adapter speed is not selected in the init scripts. OpenOCD will try to run the adapter at the low speed (100 kHz)
Warn : To remove this warnings and achieve reasonable communication speed with the target, set "adapter speed" or "jtag_rclk" in the init scripts.
Info : Using CMSIS-DAPv2 interface with VID:PID=0x2e8a:0x000c, serial=E6614103E76B342F
Info : CMSIS-DAP: SWD supported
Info : CMSIS-DAP: Atomic commands supported
Info : CMSIS-DAP: Test domain timer supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 100 kHz
Info : SWD DPIDR 0x0bc12477, DLPIDR 0x00000001
Info : SWD DPIDR 0x0bc12477, DLPIDR 0x10000001
Info : [rp2040.core0] Cortex-M0+ r0p1 processor detected
Info : [rp2040.core0] target has 4 breakpoints, 2 watchpoints
Info : [rp2040.core1] Cortex-M0+ r0p1 processor detected
Info : [rp2040.core1] target has 4 breakpoints, 2 watchpoints
Info : starting gdb server for rp2040.core0 on 3333
Info : Listening on port 3333 for gdb connections
Error: error reading data: Pipe error
Error: CMSIS-DAP command CMD_DISCONNECT failed.
Error: CMSIS-DAP command CMD_CONNECT failed.
Error: CMSIS-DAP command SWD_Sequence failed.
```

**Reply by maxgerhardt** [reply]
Great, now it has recognized both cortex m0+ cores of your rp2040 target. Do you connect to it with GDB afterwards or when does the disconnect error appear?

What's the high-level goal here, debugging a .elf firmware on the commandline with GDB or debugging a firmware in an IDE like VSCode?

**Comment by approxdec** [comment]
thanks.
i am getting openocd connecting to gdb
i am rereading the getting started doc for connect to gdb sequence

initially i will get th commandline working and then try for VSCode

**Reply by approxdec** [reply]
following the sequence in the getting started doc
connecting to  gdb

```
Remote debugging using localhost:3333
0x000001ae in ?? ()
(gdb) l
No symbol table is loaded.  Use the "file" command.
```
i have installed the windows toolchain include VSCode and the examples
built and run the examples in VSCode and on the command line
but that has not built the debug version of say "Hello_World"
so now i need to figure out how to do that from the command line

but thank you
it seems i have the debug probe connecting to a pico
openocd running, connecting gdb

i'm using openocd from https://github.com/raspberrypi/pico-setup-windows/releases/latest/download/openocd-x64-standalone.zip - that is the link suggested for Windows in the Raspberry Pi Debug Probe doc
and GDB (in the pqt-gcc directory) from the arduin-pico package

once things are settled i will try OpenOCD (in the pqt-openocd directory)

okay

back to the debug building

**Reply by approxdec** [reply]
debug build finished
[with slight changes for windows
openocd + gdb = connected as in the getting started doc

crikey. thank very much Max for pointing the way for me.
i will review everything we have done and have a little drink in your honour.

i have set up the windows/VSCode environment but i love the arduino-pico package
and you have opened up the Raspberry Pi Debug Probe -openocd-gdb for me.

thanks again. when i have settled everything in my head i will try OpenOCD (in the pqt-openocd directory)
and let you know how that goes.

regards

joe

**Reply by maxgerhardt** [reply]
If you wanted a more seamless experience of VSCode + Arduino-Pico + Debugging you could also just use [PlatformIO](https://arduino-pico.readthedocs.io/en/latest/platformio.html#debugging) and chose the `debug_tool = cmsis-dap`.
